CRM/Price/BAO/PriceSet.php

```diff
@@ -1672,6 +1669,236 @@ public static function getNonDeductibleAmountFromPriceSet($priceSetId, $lineItem
     }
 
     return $nonDeductibleAmount;
+
+  }
+
+  /**
+   * Get the last (still active) price set that was used to create that membership.
+   *
+   * If that price set does not exist, is not found or isn't active,
+   * then return the price set that will  cover that membership's record organisation.
+   * As multiple price sets may meet that criteria, the price set that cover's the most
+   * of that contact's membership of organizations will be returned.  As multiple price sets may
+   * still meet that criteria, the first defined price set is returned. Getting
+   * heuristics for determining the best price set here seems impossible.
+   *
+   * @param int $membership_id The membership that is being renewed.
+   *
+   * @return int
+   *     e.g., 9. To indicate that price set #9 is the last price set that as
+   *           used on contact #3.
+   */
+  public static function getLastPriceSetUsed($membership_id) {
+    $sql = "
+SELECT p.contribution_id, c.receive_Date, pf.price_set_id
+  FROM civicrm_membership m
+    LEFT JOIN civicrm_membership_payment p
+           ON p.membership_id = m.id
+     LEFT JOIN civicrm_contribution c
+            ON c.id = p.contribution_id
+    LEFT JOIN civicrm_line_item li
+           ON li.contribution_id = p.contribution_id
+          AND li.entity_table = 'civicrm_membership'
+          AND li.entity_id = m.id
+    LEFT JOIN civicrm_price_field_value pfv
+           ON pfv.id = li.price_field_value_id
+          AND pfv.is_active = 1
+    LEFT JOIN civicrm_price_field pf
+           ON pf.id = pfv.price_field_id
+          AND pf.is_active = 1
+    LEFT JOIN civicrm_price_set ps
+           ON ps.id = pf.price_set_id
+          AND ps.is_active = 1
+ WHERE m.id = %1
+       ORDER BY c.receive_date DESC, p.contribution_id DESC
+    ";
+
+    $params = array(1 => array($membership_id, 'Integer'));
+    $dao = self::executeQuery($sql, $params);
+    $activateByDefault = TRUE;
+    while ($dao->fetch()) {
+      // last price set used on this membership (could be multiple over the course of time)
+      // only select "price set view", if last contribution was for the price set.
+      if ($dao->price_set_id) {
+        return array(
+          'price_set_id' => $dao->price_set_id,
+          'price_set_is_through_contribution' => $activateByDefault,
+        );
+      }
+      $activateByDefault = FALSE;
+    }
+
+    // nothing found.  we want to pick, by default the price set that will cover
```

Comment:
The business goal was to always default the best possible default.  So the requirement that I set was this.
1) Use last price set used for membership being renewed (this one seems evident).
2) If that doesn't work, pick a price set that will match all of this contact's memberships

In the case of my client, #2 worked well because of how their data was converted: initial memberships were imported without contributions, so it would be impossible to determine a priceset for those initial conversions.

Objectively, not sure many people would care much about this second rule, specially since it only comes into play after user clicks on "click here" from [this portion](https://issues.civicrm.org/jira/secure/attachment/53277/53277_Screen+Shot+2.png) of the renew membership dialog.

That being said, this would not return a different default depending on time of year/membership evolution (barring, price set config changes and/or deletions/additions of new membership to that contact's actual memberships -- but then all bets are off).

It would be possible, however, to get a different default depending on contact selected, potentially of concern, but also explainable with above rule. Weighing pros (a default in an edge case) vs cons (added complexity + potentially some explainable confusion on default), I'm almost talking myself out of it.

So I'm fine removing this altogether.  Would just need to explain to client (which they'll probably be fine with) and/or add an extension hook spot here.